From 04b23a9251dcce9476c8db8be211aefae59c657a Mon Sep 17 00:00:00 2001
From: Benjamin Cheng <benjamin.cheng@amd.com>
Date: Mon, 15 Jun 2026 16:44:59 -0400
Subject: [PATCH] mesa/st: use col0 attrib from provoking vertex for feedback

When clipping triangles and flatshade is enabled, the clip stage does
not copy the col0 attribute to the new vertices. GL_FEEDBACK does not
expect flatshaded outputs, and thus uninitialized memory is returned to
the app.

Handle it in the st_cb_feedback by copying col0 from the provoking vertex.

Cc: mesa-stable
Closes: https://gitlab.freedesktop.org/mesa/mesa/-/work_items/15304
Signed-off-by: Benjamin Cheng <benjamin.cheng@amd.com>
Reviewed-by: Roland Scheidegger <roland.scheidegger@broadcom.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/41441>
---
 src/mesa/state_tracker/st_cb_feedback.c | 20 ++++++++++++--------
 1 file changed, 12 insertions(+), 8 deletions(-)

diff --git a/src/mesa/state_tracker/st_cb_feedback.c b/src/mesa/state_tracker/st_cb_feedback.c
index a69c93eaa47e..c5757eb4cb9b 100644
--- a/src/mesa/state_tracker/st_cb_feedback.c
+++ b/src/mesa/state_tracker/st_cb_feedback.c
@@ -84,7 +84,7 @@ feedback_stage( struct draw_stage *stage )
 
 static void
 feedback_vertex(struct gl_context *ctx, const struct draw_context *draw,
-                const struct vertex_header *v)
+                const struct vertex_header *v, const struct vertex_header *pv)
 {
    struct gl_vertex_program *stvp =
       (struct gl_vertex_program *)ctx->VertexProgram._Current;
@@ -107,7 +107,7 @@ feedback_vertex(struct gl_context *ctx, const struct draw_context *draw,
 
    slot = stvp->result_to_output[VARYING_SLOT_COL0];
    if (slot != 0xff)
-      color = v->data[slot];
+      color = pv ? pv->data[slot] : v->data[slot];
    else
       color = ctx->Current.Attrib[VERT_ATTRIB_COLOR0];
 
@@ -126,11 +126,15 @@ feedback_tri( struct draw_stage *stage, struct prim_header *prim )
 {
    struct feedback_stage *fs = feedback_stage(stage);
    struct draw_context *draw = stage->draw;
+   const struct vertex_header *pv = NULL;
+   if (draw->rasterizer->flatshade)
+      pv = draw->rasterizer->flatshade_first ? prim->v[0] : prim->v[2];
+
    _mesa_feedback_token(fs->ctx, (GLfloat) GL_POLYGON_TOKEN);
    _mesa_feedback_token(fs->ctx, (GLfloat) 3); /* three vertices */
-   feedback_vertex(fs->ctx, draw, prim->v[0]);
-   feedback_vertex(fs->ctx, draw, prim->v[1]);
-   feedback_vertex(fs->ctx, draw, prim->v[2]);
+   feedback_vertex(fs->ctx, draw, prim->v[0], pv);
+   feedback_vertex(fs->ctx, draw, prim->v[1], pv);
+   feedback_vertex(fs->ctx, draw, prim->v[2], pv);
 }
 
 
@@ -146,8 +150,8 @@ feedback_line( struct draw_stage *stage, struct prim_header *prim )
    else {
       _mesa_feedback_token(fs->ctx, (GLfloat) GL_LINE_TOKEN);
    }
-   feedback_vertex(fs->ctx, draw, prim->v[0]);
-   feedback_vertex(fs->ctx, draw, prim->v[1]);
+   feedback_vertex(fs->ctx, draw, prim->v[0], NULL);
+   feedback_vertex(fs->ctx, draw, prim->v[1], NULL);
 }
 
 
@@ -157,7 +161,7 @@ feedback_point( struct draw_stage *stage, struct prim_header *prim )
    struct feedback_stage *fs = feedback_stage(stage);
    struct draw_context *draw = stage->draw;
    _mesa_feedback_token(fs->ctx, (GLfloat) GL_POINT_TOKEN);
-   feedback_vertex(fs->ctx, draw, prim->v[0]);
+   feedback_vertex(fs->ctx, draw, prim->v[0], NULL);
 }
 
 
-- 
GitLab
